compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -3172,6 +3180,127 @@ private STNode parseActionOrExpressionInLhs(SyntaxKind nextTokenKind, STNode lhs
+     * <p>
+     * Parse `new` keyword.
+     * </p>
+     *
+     * @return Parsed NEW_KEYWORD Token.
+     */
+    private STNode parseNewKeyword() {
+        STToken token = peek();
+        if (token.kind == SyntaxKind.NEW_KEYWORD) {
+            return consume();
+        } else {
+            Solution sol = recover(token, ParserRuleContext.NEW_KEYWORD);
+            return sol.recoveredNode;
+        }
+    }
+
+    private STNode parseNewExpression(STNode newKeyword) {
+        STNode token = peek();
+        return parseNewExpression(token.kind, newKeyword);
+    }
+
+    /**
+     * <p>
+     * Parse an implicit or explicit expression.
+     * </p>
+     * @param kind next token kind.
+     * @param newKeyword parsed node for `new` keyword.
+     * @return Parsed new-expression node.
+     */
+    private STNode parseNewExpression(SyntaxKind kind, STNode newKeyword) {
+        switch (kind) {
+            case OPEN_PAREN_TOKEN:
+                return parseImplicitNewExpression(newKeyword);
+            case SEMICOLON_TOKEN:
+                return STNodeFactory.createImplicitNewExpression(newKeyword, STNodeFactory.createEmptyNode());
+            case IDENTIFIER_TOKEN:
+            case OBJECT_KEYWORD:
+                // TODO: Support `stream` keyword once introduced
+                return parseExplicitNewExpression(newKeyword);
+            default:
+                Solution sol = recover(peek(), ParserRuleContext.NEW_EXPRESSION, newKeyword);
+                return parseNewExpression(sol.recoveredNode.kind, newKeyword);
+        }
+    }
+
+    /**
+     * <p>
+     * Parse an Explicit New expression.
+     * </p>
+     * <code>
+     *  explicit-new-expr := new type-descriptor ( arg-list )
+     * </code>
+     *
+     * @param newKeyword Parsed `new` keyword.
+     * @return the Parsed Explicit New Expression.
+     */
+    private STNode parseExplicitNewExpression(STNode newKeyword) {
+        startContext(ParserRuleContext.EXPLICIT_NEW_RHS);
+        STNode typeDescriptor = parseTypeDescriptor();
+        STNode parenthesizedArgsList = parseParenthesizedArgList();
+        endContext();
+
+        return STNodeFactory.createExplicitNewExpression(newKeyword, typeDescriptor, parenthesizedArgsList);
+    }
+
+    /**
+     * <p>
+     * Parse an <code>implicit-new-expr</code> with arguments.
+     * </p>
+     *
+     * @param newKeyword Parsed `new` keyword.
+     * @return Parsed implicit-new-expr.
+     */
+    private STNode parseImplicitNewExpression(STNode newKeyword) {
+        STNode implicitNewArgList = parseParenthesizedArgList();
+
+        return STNodeFactory.createImplicitNewExpression(newKeyword, implicitNewArgList);
+    }
+
+    /**
+     * <p>
+     * Parse the parenthesized argument list for a <code>new-expr</code>.
+     * </p>
+     *
+     * @return Parsed parenthesized rhs of <code>new-expr</code>.
+     */
+    private STNode parseParenthesizedArgList() {
+        startContext(ParserRuleContext.NEW_RHS);
```

Comment:
Shall we have a different name instead of `NEW_RHS`. When we say new-rhs, what comes to mind is all possible alternatives after `new` keyword. i.e: `(open-paren | identifier | expr-rhs)`

Reply:
I'll use new-expr-args? @SupunS

Reply:
fixed with 07c06fd using new-expr-args. Let me know if I should use something else